File: python/poker/poker.py
import collections

# Card ranks mapping (T=10, J=11, Q=12, K=13, A=14)
# Ace can also be low (1) for A-2-3-4-5 straights, handled separately.
RANKS = {str(n): n for n in range(2, 10)}
RANKS.update({'T': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14})
RANK_ORDER = '23456789TJQKA'

# Hand ranks (higher is better)
HIGH_CARD = 0
ONE_PAIR = 1
TWO_PAIR = 2
THREE_OF_A_KIND = 3
STRAIGHT = 4
FLUSH = 5
FULL_HOUSE = 6
FOUR_OF_A_KIND = 7
STRAIGHT_FLUSH = 8

def parse_card(card_str):
    """Parses a card string like '4S' into (rank, suit)."""
    if len(card_str) == 3: # Handle '10' rank
        rank_str = 'T' # Use 'T' for rank 10
        suit = card_str[2]
    else:
        rank_str = card_str[0]
        suit = card_str[1]
    rank = RANKS[rank_str]
    return rank, suit

def score_hand(hand_str):
    """Scores a poker hand string. Returns a tuple for comparison.
    Format: (hand_rank, kicker_ranks...)
    Higher hand_rank wins. If tied, compare kickers in order.
    """
    cards = sorted([parse_card(c) for c in hand_str.split()], key=lambda x: x[0], reverse=True)
    ranks = [card[0] for card in cards]
    suits = [card[1] for card in cards]

    is_flush = len(set(suits)) == 1
    # Check for A-2-3-4-5 straight (ranks are 14, 5, 4, 3, 2)
    is_low_ace_straight = ranks == [14, 5, 4, 3, 2]
    # Check for normal straight
    is_straight = all(ranks[i] == ranks[0] - i for i in range(5)) or is_low_ace_straight

    # Adjust ranks for A-2-3-4-5 straight for comparison (treat Ace as low)
    if is_low_ace_straight:
        ranks = [5, 4, 3, 2, 1] # Use 1 for Ace's rank in this specific straight

    # Straight Flush
    if is_straight and is_flush:
        # Highest card determines rank (use adjusted ranks for A-5 straight)
        return (STRAIGHT_FLUSH, ranks[0])

    rank_counts = collections.Counter(ranks)
    counts = sorted(rank_counts.values(), reverse=True)
    most_common_ranks = sorted(rank_counts.keys(), key=lambda r: (rank_counts[r], r), reverse=True)

    # Four of a Kind
    if counts == [4, 1]:
        four_rank = most_common_ranks[0]
        kicker = most_common_ranks[1]
        return (FOUR_OF_A_KIND, four_rank, kicker)

    # Full House
    if counts == [3, 2]:
        three_rank = most_common_ranks[0]
        pair_rank = most_common_ranks[1]
        return (FULL_HOUSE, three_rank, pair_rank)

    # Flush
    if is_flush:
        return (FLUSH,) + tuple(ranks)

    # Straight
    if is_straight:
         # Highest card determines rank (use adjusted ranks for A-5 straight)
        return (STRAIGHT, ranks[0])

    # Three of a Kind
    if counts == [3, 1, 1]:
        three_rank = most_common_ranks[0]
        kickers = sorted([r for r in ranks if r != three_rank], reverse=True)
        return (THREE_OF_A_KIND, three_rank) + tuple(kickers)

    # Two Pair
    if counts == [2, 2, 1]:
        pair_ranks = sorted([r for r in most_common_ranks if rank_counts[r] == 2], reverse=True)
        kicker = most_common_ranks[2]
        return (TWO_PAIR,) + tuple(pair_ranks) + (kicker,)

    # One Pair
    if counts == [2, 1, 1, 1]:
        pair_rank = most_common_ranks[0]
        kickers = sorted([r for r in ranks if r != pair_rank], reverse=True)
        return (ONE_PAIR, pair_rank) + tuple(kickers)

    # High Card
    return (HIGH_CARD,) + tuple(ranks)


def best_hands(hands):
    """Given a list of poker hands, return the best hand(s)."""
    if not hands:
        return []
    if len(hands) == 1:
        return hands

    scored_hands = []
    for hand_str in hands:
        score = score_hand(hand_str)
        scored_hands.append((score, hand_str))

    # Find the maximum score
    max_score = max(scored_hands, key=lambda x: x[0])[0]

    # Return all hands that match the maximum score
    winners = [hand_str for score, hand_str in scored_hands if score == max_score]
    return winners

File: java/poker/src/main/java/Poker.java
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class Poker {
    private final List<Hand> hands;
    private final List<Hand> bestHands;

    Poker(List<String> handStrings) {
        this.hands = handStrings.stream().map(Hand::new).collect(Collectors.toList());
        this.bestHands = findBestHands();
    }

    List<String> getBestHands() {
        return bestHands.stream().map(Hand::toString).collect(Collectors.toList());
    }

    private List<Hand> findBestHands() {
        if (hands.isEmpty()) {
            return Collections.emptyList();
        }
        List<Hand> sortedHands = new ArrayList<>(hands);
        sortedHands.sort(Collections.reverseOrder()); // Sort descending, best hands first
        Hand bestHand = sortedHands.get(0);
        return sortedHands.stream()
                .filter(hand -> hand.compareTo(bestHand) == 0)
                .collect(Collectors.toList());
    }

    // --- Inner Classes ---

    enum Suit { S, H, D, C }

    enum HandRank {
        HIGH_CARD, ONE_PAIR, TWO_PAIR, THREE_OF_A_KIND, STRAIGHT,
        FLUSH, FULL_HOUSE, FOUR_OF_A_KIND, STRAIGHT_FLUSH
    }

    static class Card implements Comparable<Card> {
        final int rank; // 2-10, J=11, Q=12, K=13, A=14
        final Suit suit;
        final String representation; // e.g., "KH", "AS"

        Card(String cardStr) {
            this.representation = cardStr;
            char suitChar = cardStr.charAt(cardStr.length() - 1);
            String rankStr = cardStr.substring(0, cardStr.length() - 1);

            this.rank = switch (rankStr) {
                case "T" -> 10;
                case "J" -> 11;
                case "Q" -> 12;
                case "K" -> 13;
                case "A" -> 14;
                default -> Integer.parseInt(rankStr); // Handles 2-10
            };

            this.suit = switch (suitChar) {
                case 'S' -> Suit.S;
                case 'H' -> Suit.H;
                case 'D' -> Suit.D;
                case 'C' -> Suit.C;
                default -> throw new IllegalArgumentException("Invalid suit: " + suitChar);
            };
        }

        @Override
        public int compareTo(Card other) {
            return Integer.compare(this.rank, other.rank);
        }

        @Override
        public String toString() {
            return representation;
        }
    }

    static class Hand implements Comparable<Hand> {
        final List<Card> cards;
        final String originalString;
        final HandRank handRank;
        final List<Integer> tieBreakingRanks; // Ranks used for tie-breaking, highest first

        Hand(String handStr) {
            this.originalString = handStr;
            this.cards = Arrays.stream(handStr.split(" "))
                    .map(Card::new)
                    .sorted(Comparator.comparingInt((Card c) -> c.rank).reversed()) // Sort cards high to low
                    .collect(Collectors.toList());

            if (this.cards.size() != 5) {
                throw new IllegalArgumentException("Hand must contain 5 cards: " + handStr);
            }

            EvaluationResult eval = evaluateHand();
            this.handRank = eval.rank;
            this.tieBreakingRanks = eval.tieBreakingRanks;
        }

        private EvaluationResult evaluateHand() {
            boolean isFlush = cards.stream().map(c -> c.suit).distinct().count() == 1;
            boolean isStraight = isStraight(cards);

            if (isStraight && isFlush) {
                // Check for Ace-low straight flush (A 2 3 4 5)
                if (cards.get(0).rank == 14 && cards.get(1).rank == 5) {
                     return new EvaluationResult(HandRank.STRAIGHT_FLUSH, List.of(5)); // A-5 uses 5 as high card
                }
                return new EvaluationResult(HandRank.STRAIGHT_FLUSH, List.of(cards.get(0).rank));
            }

            Map<Integer, Long> rankCounts = cards.stream()
                    .collect(Collectors.groupingBy(c -> c.rank, Collectors.counting()));

            List<Integer> counts = rankCounts.values().stream().sorted(Comparator.reverseOrder()).map(Long::intValue).collect(Collectors.toList());
            List<Integer> ranksByFrequency = rankCounts.entrySet().stream()
                    .sorted(Map.Entry.<Integer, Long>comparingByValue().reversed()
                            .thenComparing(Map.Entry.comparingByKey(Comparator.reverseOrder())))
                    .map(Map.Entry::getKey)
                    .collect(Collectors.toList());

            if (counts.get(0) == 4) { // Four of a Kind
                return new EvaluationResult(HandRank.FOUR_OF_A_KIND, ranksByFrequency); // [Rank of 4, Kicker]
            }
            if (counts.get(0) == 3 && counts.get(1) == 2) { // Full House
                return new EvaluationResult(HandRank.FULL_HOUSE, ranksByFrequency); // [Rank of 3, Rank of 2]
            }
            if (isFlush) {
                return new EvaluationResult(HandRank.FLUSH, cards.stream().map(c -> c.rank).collect(Collectors.toList())); // All card ranks
            }
             if (isStraight) {
                // Check for Ace-low straight (A 2 3 4 5)
                if (cards.get(0).rank == 14 && cards.get(1).rank == 5) {
                     return new EvaluationResult(HandRank.STRAIGHT, List.of(5)); // A-5 uses 5 as high card
                }
                return new EvaluationResult(HandRank.STRAIGHT, List.of(cards.get(0).rank));
            }
            if (counts.get(0) == 3) { // Three of a Kind
                return new EvaluationResult(HandRank.THREE_OF_A_KIND, ranksByFrequency); // [Rank of 3, Kicker1, Kicker2]
            }
            if (counts.get(0) == 2 && counts.get(1) == 2) { // Two Pair
                return new EvaluationResult(HandRank.TWO_PAIR, ranksByFrequency); // [High Pair Rank, Low Pair Rank, Kicker]
            }
            if (counts.get(0) == 2) { // One Pair
                return new EvaluationResult(HandRank.ONE_PAIR, ranksByFrequency); // [Pair Rank, Kicker1, Kicker2, Kicker3]
            }
            // High Card
            return new EvaluationResult(HandRank.HIGH_CARD, cards.stream().map(c -> c.rank).collect(Collectors.toList())); // All card ranks
        }

        private boolean isStraight(List<Card> sortedCards) {
            // Check for Ace-low straight (A 2 3 4 5)
            boolean aceLowStraight = sortedCards.get(0).rank == 14 &&
                                     sortedCards.get(1).rank == 5 &&
                                     sortedCards.get(2).rank == 4 &&
                                     sortedCards.get(3).rank == 3 &&
                                     sortedCards.get(4).rank == 2;
            if (aceLowStraight) return true;

            // Check for regular straight
            return IntStream.range(0, sortedCards.size() - 1)
                    .allMatch(i -> sortedCards.get(i).rank == sortedCards.get(i + 1).rank + 1);
        }


        @Override
        public int compareTo(Hand other) {
            int rankComparison = this.handRank.compareTo(other.handRank);
            if (rankComparison != 0) {
                return rankComparison;
            }
            // Tie-breaking based on ranks
            for (int i = 0; i < this.tieBreakingRanks.size(); i++) {
                int tieComparison = Integer.compare(this.tieBreakingRanks.get(i), other.tieBreakingRanks.get(i));
                if (tieComparison != 0) {
                    return tieComparison;
                }
            }
            return 0; // Hands are exactly equal
        }

        @Override
        public String toString() {
            return originalString;
        }

        private record EvaluationResult(HandRank rank, List<Integer> tieBreakingRanks) {}
    }
}